### PR TITLE
refactor: publish post funding application events from resource class…

### PR DESCRIPTION
**Pull request by CaiWillis**
… as temporary workaround for transaction timing issues

TIS21-5811: Update fundingStatus once a post funding is created/updated/deleted

### CODE REVIEW

File: tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostFundingResourceTest.java
Comment (JosephKelly) [inline tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostFundingResourceTest.java:1]: I personally prefer integrating the verification with the tests in `PostFundingResourceIntTest.java`.  See examples with `ArgumentCaptor`s in [PostServiceImplTest](https://github.com/Health-Education-England/TIS-TCS/blob/703ac3dc7e49558b8618179d8ac82bb4bd8423d3/tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/service/impl/PostServiceImplTest.java#L209)

File: tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostResource2Test.java [hunk 1]
```diff
@@ -260,6 +289,20 @@ public void updatePostShouldReturnUnauth() throws Exception {
         .andExpect(jsonPath("$.message").value("error.accessDenied"));
   }
 
+  @Test
+  public void updatePostShouldPublishPostSavedEvent() throws Exception {
```
Comment (JosephKelly) [inline tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostResource2Test.java:293]: test name suggests it should verify that an appropriate `PostSavedEvent` was published but instead the HTTP Response is verified. `createPostShouldPublishPostSavedEvent()` verifies both and I'm not sure that's necessary / a useful addition.

**Review by JosephKelly — CHANGES_REQUESTED** [review]
Please will you review the tests?  There is a balance to be struck between getting really specific feedback from separate tests, each verifying different aspects and the additional build time caused by duplication.

File: tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostResource2Test.java (hunk 1 above)
Comment (CaiWillis) [reply tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostResource2Test.java:293]: Ah this is just a test I didn't finish

**Review by CaiWillis — COMMENTED** [review]
Flyway script to set values correctly - previous commit has been pushed to prod
UPDATE: Done in latest commit

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/PostFundingResource.java [hunk 2]
```diff
@@ -43,11 +47,21 @@ public class PostFundingResource {
...
   private final PostFundingValidator postFundingValidator;
+  private final ApplicationEventPublisher applicationEventPublisher;
 
+  /**
+   * Constructor.
+   *
+   * @param postFundingService        service
+   * @param postFundingValidator      mapper
+   * @param applicationEventPublisher event publisher
+   */
```
Comment (JosephKelly) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/PostFundingResource.java:58]: The code smells are there to encourage good practice.  We should go beyond satisfying the computer validated requirement.  Also the validator isn't a mapper :wink:.

 ```java
  /**
   * Constructs a PostFundingResource with required dependencies.
   *
   * @param postFundingService        encapsulates operations including persistence
   * @param postFundingValidator      validates {@link PostFundingDTO}s, including reference data
   * @param applicationEventPublisher publishes events for changes to a PostFunding
   */
```

Comment (CaiWillis) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/PostFundingResource.java:58]: That'll teach me to think I can copy and paste and get away with it 😁

**Review by JosephKelly — APPROVED** [review]

**Pull request merged**